README.md
# @npms/queries

[![NPM version][npm-image]][npm-url] [![Downloads][downloads-image]][npm-url] [![Build Status][travis-image]][travis-url] [![Coverage status][codecov-image]][codecov-url] [![Dependency status][david-dm-image]][david-dm-url] [![Dev Dependency status][david-dm-dev-image]][david-dm-dev-url]

Module that offers a variety of queries around `npms` data.


## Installation

`$ npm install @npms/queries`


### Usage

For now, only queries related to `search` are available. Though, the goal of this module is to provide other interesting queries in the near future, such as top ranked modules, top authors, etc.


#### .search(q, esClient, [options]) -> Promise

Performs a search query.

Besides normal text, `q` supports qualifiers to express filters and other modifiers.
The `esClient` accepts a [`elasticsearch`](https://github.com/elastic/elasticsearch-js) instance or a config to instantiate it.

You may read the [API docs](https://api-docs.npms.io/#api-Search-ExecuteSearchQuery) to know which qualifiers are available.

```js
const queries = require('@npms/queries');

// ...
queries.search('test framework', esClient)
.then((res) => {
    console.log('total', res.total);
    console.log('results', res.results);
});
```

Available options:

- `from`: The offset in which to start searching from, defaults to `0`
- `size`: The total number of results to return, defaults to `25`
- `throwOnInvalid`: Whether to reject the promise if the query has invalid qualifiers or not, defaults to `false` (if `false`, invalid qualifiers will be removed from `q`)


#### .search.suggestions(q, esClient, [options]) -> Promise

Fetch search suggestions to be typically displayed when doing autocomplete.

Only normal text is supported in `q` but any qualifiers will be automatically discarded.
The `esClient` accepts a [`elasticsearch`](https://github.com/elastic/elasticsearch-js) instance or a config to instantiate it.

```js
const queries = require('@npms/queries');

// ...
queries.search.suggestions('gulp', esClient)
.then((suggestions) => console.log('suggestions', suggestions));
```

Available options:

- `boostExact`: How much should the score of exact matches be boosted? defaults to `100000`.
- `size`: The total number of results to return, defaults to `25`
- `analyzerWeight`: How much should we weight the analyzer's `score.final` by? defaults to `1.0`.
- `scoreWeight`: How much should we weight the search `_score`? defaults to `0.3`.

#### .search.similar(q, esClient, [options]) -> Promise

Perform a fuzzy search for similarly named packages.

Results are ranked based on a combination of analyzer weightings (`quality`, `popularity`, `maintenance`) and the `_score` returned by the [fuzzy match](https://www.elastic.co/guide/en/elasticsearch/guide/current/fuzzy-match-query.html).

```js
const queries = require('@npms/queries');

// ...
queries.search.similar('chaik', esClient)
.then(results => {
  // perhaps we were instead looking for chalk?
});
```

Available options:

- `size`: The total number of results to return, defaults to `10`.
- `analyzerWeight`: How much should we weight the analyzer's `score.final` by? defaults to `2.2`.
- `scoreWeight`: How much should we weight the search `_score`? defaults to `1.5`.
- `minScore`: defaults to `4.5`.

_the above default values were based on trial and error examining the
 top npm modules, they will likely change over time._

## Tests

`$ npm test`   
`$ npm test-cov` to get coverage report


## License

Released under the [MIT License](http://www.opensource.org/licenses/mit-license.php).


[npm-url]:https://npmjs.org/package/%40npms%2Fqueries
[downloads-image]:http://img.shields.io/npm/dm/%40npms%2Fqueries.svg
[npm-image]:http://img.shields.io/npm/v/%40npms%2Fqueries.svg
[travis-url]:https://travis-ci.org/npms-io/queries
[travis-image]:http://img.shields.io/travis/npms-io/queries/master.svg
[codecov-url]:https://codecov.io/gh/npms-io/npms-badges
[codecov-image]:https://img.shields.io/codecov/c/github/npms-io/npms-badges/master.svg
[david-dm-url]:https://david-dm.org/npms-io/queries
[david-dm-image]:https://img.shields.io/david/npms-io/queries.svg
[david-dm-dev-url]:https://david-dm.org/npms-io/queries#info=devDependencies
[david-dm-dev-image]:https://img.shields.io/david/dev/npms-io/queries.svg


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.0.0"></a>
# [3.0.0](https://github.com/npms-io/queries/compare/v2.0.2...v3.0.0) (2018-10-30)


### Features

* upgrade to new index & elasticsearch ([7465d37](https://github.com/npms-io/queries/commit/7465d37))


### BREAKING CHANGES

* the new queries are incompatible with the old npms index



<a name="2.0.2"></a>
## [2.0.2](https://github.com/npms-io/queries/compare/v2.0.1...v2.0.2) (2017-11-09)


### Bug Fixes

* boolean operation needed parenthesis in groovy ([6dce5ac](https://github.com/npms-io/queries/commit/6dce5ac))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/npms-io/queries/compare/v2.0.0...v2.0.1) (2017-11-09)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/npms-io/queries/compare/v1.3.0...v2.0.0) (2017-11-09)


### Features

* take into account the analyzer's score.final when sorting suggestions ([#11](https://github.com/npms-io/queries/issues/11)) ([c8b1297](https://github.com/npms-io/queries/commit/c8b1297))


### BREAKING CHANGES

* analyzer score now affects typeahead results


test/test.js
'use strict';

const nock = require('nock');
const nockBack = require('nock').back;

beforeEach(() => nock.cleanAll());
nockBack.fixtures = `${__dirname}/fixtures`;

require('./spec/util/parseSearchQuery');
require('./spec/search');
require('./spec/searchSuggestions');
require('./spec/searchSimilar');


test/spec/search.js
'use strict';

const expect = require('chai').expect;
const elasticsearch = require('elasticsearch');
const nockBack = require('nock').back;
const queries = require('../../');

const localEsClient = new elasticsearch.Client({ host: '127.0.0.1:9200', log: null, apiVersion: '6.3' });

describe('search()', () => {
    it('should return the desired results', () => {
        let nockDone;

        nockBack('search-gulpplugin-sass.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('keywords:gulpplugin sass', localEsClient)
        .then((res) => {
            expect(res.total).to.equal(108);
            expect(res.results).to.have.length(25);
            expect(res.results[0]).to.contain.all.keys('package', 'score', 'searchScore');
            expect(res.results[0].package.name).to.equal('gulp-sass');
            nockDone();
        });
    });

    it('should allow to search all packages of an author', () => {
        let nockDone;

        nockBack('search-author-sindresorhus.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('author:sindresorhus', localEsClient)
        .then((res) => {
            expect(res.total).to.equal(943);
            expect(res.results).to.have.length(25);
            res.results.forEach((result) => expect(result.package.author.username).to.equal('sindresorhus'));
            nockDone();
        });
    });

    it('should allow to search all packages by scope', () => {
        let nockDone;

        nockBack('search-scope-bcoe.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('scope:bcoe', localEsClient)
        .then((res) => {
            expect(res.results).to.have.length(17);
            res.results.forEach((result) => expect(result.package.scope).to.equal('bcoe'));
            nockDone();
        });
    });

    it('should make use of options.from and options.size', () => {
        let nockDone;

        nockBack('search-react-with-from-and-limit.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('react', localEsClient, { from: 1, size: 1 })
        .then((res) => {
            expect(res.total).to.be.greaterThan(1);
            expect(res.results).to.have.length(1);
            expect(res.results[0].package.name).to.not.equal('react');
            nockDone();
        });
    });

    it('should return the desired results of a complex query', () => {
        let nockDone;

        nockBack('search-complex-query-spawn.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('maintainer:satazor keywords:spawn,-foo is:deprecated not:insecure boost-exact:no spawn', localEsClient)
        .then((res) => {
            expect(res.total).to.equal(1);
            expect(res.results).to.have.length(1);
            expect(res.results[0].package.name).to.equal('cross-spawn-async');
            nockDone();
        });
    });

    it('should accept a elasticsearch config as the second argument', () => {
        let nockDone;

        nockBack('search-gulpplugin-sass.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('keywords:gulpplugin sass', { host: '127.0.0.1:9200', log: null, apiVersion: '6.3' })
        .then((res) => {
            expect(res.total).to.equal(108);
            nockDone();
        });
    });

    it('should not fail on invalid qualifiers by default', () => {
        let nockDone;

        nockBack('search-gulpplugin-sass.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search('keywords:gulpplugin is:foo sass', localEsClient)
        .then((res) => {
            expect(res.total).to.equal(108);
            nockDone();
        });
    });

    it('should fail if q has invalid qualifiers and options.throwOnInvalid is enabled', () => (
        queries.search('keywords:gulpplugin is:foo sass', localEsClient, { throwOnInvalid: true })
        .then(() => {
            throw new Error('Should have failed');
        }, (err) => {
            expect(err).to.be.an.instanceOf(Error);
            expect(err.message).to.contain('deprecated, unstable, insecure');
        })
    ));

    it('should return 0 results if no text or filter qualifiers were specified', () => (
        queries.search('score-effect:1 quality-weight:1 popularity-weight:1 maintenance-weight:1 boost-exact:false', localEsClient)
        .then((res) => expect(res).to.eql({ total: 0, results: [] }))
    ));
});


test/spec/searchSimilar.js
'use strict';

const expect = require('chai').expect;
const elasticsearch = require('elasticsearch');
const nockBack = require('nock').back;
const queries = require('../../');

const localEsClient = new elasticsearch.Client({ host: '127.0.0.1:9200', log: null, apiVersion: '6.3' });

describe('search.similar()', () => {
    it('should return the desired results', () => {
        let nockDone;

        nockBack('search-similar.json', (_nockDone) => { nockDone = _nockDone; });

        return queries.search.similar('chaik', localEsClient)
        .then((similar) => {
            expect(similar).to.have.length(10);
            expect(similar[0]).to.contain.all.keys('package', 'score', 'searchScore');
            expect(similar[0].package.name).to.equal('chalk');
            nockDone();
        });
    });

    it('should return empty if searched without any text', () => (
        queries.search.similar('', localEsClient)
        .then((similar) => {
            expect(similar).to.have.length(0);
        })
    ));
});


lib/search.js
'use strict';

const pTry = require('p-try');
const pick = require('lodash/pick');
const parseQuery = require('./util/parseSearchQuery');
const toEsClient = require('./util/toEsClient');

function buildMatchQueries(params) {
    if (!params.text) {
        return [];
    }

    return [
        // Standard match using cross_fields
        {
            multi_match: {
                query: params.text,
                operator: 'and',
                fields: [
                    'package.name.standard^4',
                    'package.description.standard',
                    'package.keywords.standard^2',
                ],
                type: 'cross_fields',
                boost: 6,
                tie_breaker: 0.5,
            },
        },

        // Partial match using edge-ngram
        {
            multi_match: {
                query: params.text,
                operator: 'and',
                fields: [
                    'package.name.edge_ngram^4',
                    'package.description.edge_ngram',
                    'package.keywords.edge_ngram^2',
                ],
                type: 'phrase',
                slop: 3,
                boost: 3,
                tie_breaker: 0.5,
            },
        },

        // Normal term match with an english stemmer
        {
            multi_match: {
                query: params.text,
                operator: 'and',
                fields: [
                    'package.name.english_docs^4',
                    'package.description.english_docs',
                    'package.keywords.english_docs^2',
                ],
                type: 'cross_fields',
                boost: 3,
                tie_breaker: 0.5,
            },
        },

        // Normal term match with a more aggressive english stemmer (not so important)
        {
            multi_match: {
                query: params.text,
                operator: 'and',
                fields: [
                    'package.name.english_aggressive_docs^4',
                    'package.description.english_aggressive_docs',
                    'package.keywords.english_aggressive_docs^2',
                ],
                type: 'cross_fields',
                tie_breaker: 0.5,
            },
        },
    ];
}

function buildFilterQuery(params) {
    const filters = {
        must: [],
        mustNot: [],
    };

    if (params.author) {
        filters.must.push({
            bool: {
                should: [
                    { term: { 'package.author.name': params.author } },
                    { term: { 'package.author.username': params.author } },
                    { term: { 'package.author.email': params.author } },
                ],
            },
        });
    }

    if (params.maintainer) {
        filters.must.push({
            bool: {
                should: [
                    { term: { 'package.maintainers.username': params.maintainer } },
                    { term: { 'package.maintainers.email': params.maintainer } },
                ],
            },
        });
    }

    if (params.scope) {
        filters.must.push({
            term: { 'package.scope': params.scope },
        });
    }

    if (params.keywords.include.length) {
        filters.must.push({
            terms: { 'package.keywords.raw': params.keywords.include }, // Operator is OR
        });
    }
    if (params.keywords.exclude.length) {
        filters.mustNot.push({
            terms: { 'package.keywords.raw': params.keywords.exclude },
        });
    }
    if (params.is.length) {
        filters.must.push({
            bool: {
                must: params.is.map((is) => ({
                    exists: { field: `flags.${is}` },
                })),
            },
        });
    }

    if (params.not.length) {
        filters.mustNot.push({
            bool: {
                must: params.not.map((not) => ({
                    exists: { field: `flags.${not}` },
                })),
            },
        });
    }

    return filters;
}

function buildScriptScore(params) {
    let source = '(doc["score.detail.popularity"].value * params.popularityWeight + ' +
        'doc["score.detail.quality"].value * params.qualityWeight + ' +
        'doc["score.detail.maintenance"].value * params.maintenanceWeight)';

    source = params.boostExact ?
            `doc["package.name.raw"].value.equals(params.text) ? 100000 + ${source} : _score * Math.pow(${source}, params.scoreEffect)` :
            `_score * Math.pow(${source}, params.scoreEffect)`;

    return {
        script: {
            source,
            params: {
                text: params.text || '',
                scoreEffect: params.scoreEffect,
                qualityWeight: params.qualityWeight,
                popularityWeight: params.popularityWeight,
                maintenanceWeight: params.maintenanceWeight,
            },
        },
    };
}

// ------------------------------------------------------------

function search(q, esClient, options) {
    esClient = toEsClient(esClient);
    options = Object.assign({ from: 0, size: 25, throwOnInvalid: false }, options);

    return pTry(() => parseQuery(q, options))
    .then((params) => {
        if (!params) {
            return { total: 0, results: [] };
        }

        const matchQueries = buildMatchQueries(params);
        const filterQuery = buildFilterQuery(params);
        const scriptScore = buildScriptScore(params);

        // Need to use Promise.resolve because Elasticsearch does not use native Promise
        /* eslint camelcase: 0 */
        return Promise.resolve(esClient.search({
            index: 'npms-current',
            type: 'score',
            body: {
                size: options.size,
                from: options.from,
                query: {
                    function_score: {
                        boost_mode: 'replace',
                        query: {
                            bool: {
                                must: filterQuery.must,
                                must_not: filterQuery.mustNot,
                                should: matchQueries,
                                minimum_should_match: matchQueries.length ? 1 : 0,
                            },
                        },
                        script_score: scriptScore,
                    },
                },
            },
        }))
        .then((res) => ({
            total: res.hits.total,
            results: res.hits.hits.map((hit) => {
                // We can't use _fields in the query because the JSON properties order get messed up,
                // see https://github.com/elastic/elasticsearch/issues/17639
                // So we filter the source fields manually with pick().. this is not ideal since there's payload
                // navigating through the network that we do not use, but it's definitively better than having order messed up
                const result = pick(hit._source, 'package', 'flags', 'score');

                result.searchScore = hit._score;

                return result;
            }),
        }));
    });
}

module.exports = search;


lib/searchSimilar.js
'use strict';

const pick = require('lodash/pick');
const parseQuery = require('./util/parseSearchQuery');
const toEsClient = require('./util/toEsClient');

function searchSimilar(q, esClient, options) {
    esClient = toEsClient(esClient);
    options = Object.assign({
        size: 10,
        minScore: 4.5,
        analyzerWeight: 2.2,
        scoreWeight: 1.5,
    }, options);

    const text = parseQuery.discardQualifiers(q);

    if (!text) {
        return Promise.resolve([]);
    }

    const source = '(doc["score.final"].value * params.analyzerWeight) * (_score * params.scoreWeight)';

    return Promise.resolve(esClient.search({
        /* eslint camelcase: 0 */
        index: 'npms-current',
        type: 'score',
        body: {
            size: options.size,
            query: {
                function_score: {
                    min_score: options.minScore,
                    boost_mode: 'replace',
                    query: {
                        fuzzy: {
                            'package.name.raw': {
                                value: text,
                            },
                        },
                    },
                    script_score: {
                        script: {
                            source,
                            params: {
                                analyzerWeight: options.analyzerWeight,
                                scoreWeight: options.scoreWeight,
                            },
                        },
                    },
                },
            },
        },
    }))
    .then((res) => res.hits.hits.map((hit) => {
        // We can't use _fields in the query because the JSON properties order get messed up,
        // see https://github.com/elastic/elasticsearch/issues/17639
        // So we filter the source fields manually with pick().. this is not ideal since there's payload
        // navigating through the network that we do not use, but it's definitively better than having order messed up
        const result = pick(hit._source, ['package', 'flags', 'score']);

        result.searchScore = hit._score;

        return result;
    }));
}

module.exports = searchSimilar;


lib/searchSuggestions.js
'use strict';

const pick = require('lodash/pick');
const parseQuery = require('./util/parseSearchQuery');
const toEsClient = require('./util/toEsClient');

function searchSuggestions(q, esClient, options) {
    esClient = toEsClient(esClient);
    options = Object.assign({
        size: 25,
        analyzerWeight: 1,
        scoreWeight: 0.3,
        boostExact: true,
    }, options);

    const text = parseQuery.discardQualifiers(q);

    if (!text) {
        return Promise.resolve([]);
    }

    let source = '(doc["score.final"].value * params.analyzerWeight) * (_score * params.scoreWeight)';

    if (options.boostExact) {
        source += ' + (doc["package.name.raw"].value.equals(params.text) ? 1000 : 0)';
    }

    // Need to use Promise.resolve because Elasticsearch does not use native Promise
    return Promise.resolve(esClient.search({
        /* eslint camelcase: 0 */
        index: 'npms-current',
        type: 'score',
        body: {
            size: options.size,
            highlight: {
                fields: {
                    'package.name.autocomplete_highlight': {},
                },
            },
            query: {
                function_score: {
                    boost_mode: 'replace',
                    query: {
                        bool: {
                            // Match against the autocomplete field
                            must: [
                                {
                                    match_phrase: {
                                        'package.name.autocomplete': {
                                            query: text,
                                            slop: 3,
                                        },
                                    },
                                },
                            ],
                            should: [
                                // Exact prefix queries get the higher boost
                                {
                                    match: {
                                        'package.name.autocomplete_keyword': {
                                            query: text,
                                            boost: 4,
                                        },
                                    },
                                },
                                // Proximity exact match get medium boost
                                // e.g.: searching for "react form" should give higher score to "react-form-fields" than "react-formal"
                                {
                                    match_phrase: {
                                        'package.name': {
                                            query: text,
                                            slop: 0,
                                            boost: 2,
                                        },
                                    },
                                },
                                // This is just here for the highlighting
                                {
                                    match_phrase: {
                                        'package.name.autocomplete_highlight': {
                                            query: text,
                                            slop: 50, // This needs to be high because each expansion is a word
                                        },
                                    },
                                },
                            ],
                        },
                    },
                    script_score: {
                        script: {
                            source,
                            params: {
                                text,
                                scoreWeight: options.scoreWeight,
                                analyzerWeight: options.analyzerWeight,
                            },
                        },
                    },
                },
            },
        },
    }))
    .then((res) => res.hits.hits.map((hit) => {
        // We can't use _fields in the query because the JSON properties order get messed up,
        // see https://github.com/elastic/elasticsearch/issues/17639
        // So we filter the source fields manually with pick().. this is not ideal since there's payload
        // navigating through the network that we do not use, but it's definitively better than having order messed up
        const result = pick(hit._source, ['package', 'flags', 'score']);

        result.searchScore = hit._score;
        result.highlight = hit.highlight && hit.highlight['package.name.autocomplete_highlight'][0];

        return result;
    }));
}

module.exports = searchSuggestions;


lib/util/parseSearchQuery.js
'use strict';

const Joi = require('joi');
const normalizeNumber = require('normalize-number');
const searchQueryParser = require('search-query-parser');
const deepCompact = require('deep-compact');
const mapKeys = require('lodash/mapKeys');
const unset = require('lodash/unset');
const camelCase = require('lodash/camelCase');
const uniq = require('lodash/uniq');
const assignWith = require('lodash/assignWith');

const qualifiers = [
    'author', 'maintainer', 'scope', 'keywords', 'not', 'is', 'boost-exact',
    'score-effect', 'popularity-weight', 'quality-weight', 'maintenance-weight',
];

const paramsDefaults = {
    text: null,
    author: null,
    maintainer: null,
    scope: null,
    keywords: [],
    not: [],
    is: [],
    boostExact: true,
    scoreEffect: 13.5,
    qualityWeight: 1.9, // ~0.27
    popularityWeight: 3.3, // ~0.45
    maintenanceWeight: 2.1, // ~0.28
};
/* eslint-disable newline-per-chained-call */
const paramsSchema = Joi.object({
    text: Joi.string().trim().lowercase().min(1).max(250), // Trim + lowercase so that the exact boost works correctly
    author: Joi.string().trim().lowercase().min(1).max(250), // Trim + lowercase to mimic the raw analyzer
    scope: Joi.string().trim().lowercase().min(1).max(250), // Trim + lowercase to mimic the raw analyzer
    maintainer: Joi.string().trim().lowercase().min(1).max(250), // Trim + lowercase to mimic the raw analyzer
    keywords: Joi.array().items(Joi.string().trim().lowercase().min(1).max(50)).min(1).max(10).single(), // Trim + lowercase to mimic the raw analyzer
    not: Joi.array().items(Joi.string().trim().valid(['deprecated', 'unstable', 'insecure'])).single(),
    is: Joi.array().items(Joi.string().trim().valid(['deprecated', 'unstable', 'insecure'])).single(),
    boostExact: Joi.boolean().truthy('yes', '1', 1, 'on').falsy('no', '0', 0, 'off'),
    scoreEffect: Joi.number().min(0).max(25),
    qualityWeight: Joi.number().min(0).max(100),
    popularityWeight: Joi.number().min(0).max(100),
    maintenanceWeight: Joi.number().min(0).max(100),
}).required();
/* eslint-enable newline-per-chained-call */
const mandatoryParams = ['text', 'author', 'maintainer', 'scope', 'keywords', 'not', 'is'];

function normalizeWeights(params) {
    const minMax = [0, params.qualityWeight + params.popularityWeight + params.maintenanceWeight];

    params.qualityWeight = normalizeNumber(minMax, params.qualityWeight);
    params.popularityWeight = normalizeNumber(minMax, params.popularityWeight);
    params.maintenanceWeight = normalizeNumber(minMax, params.maintenanceWeight);
}

function parseKeywords(params) {
    const include = [];
    const exclude = [];

    params.keywords.forEach((keyword) => {
        if (keyword.indexOf('-') === 0) {
            exclude.push(keyword.substr(1));
        } else {
            include.push(keyword);
        }
    });

    params.keywords = { include: uniq(include), exclude: uniq(exclude) };
    params.keywords.exclude = params.keywords.exclude.filter((keyword) => params.keywords.include.indexOf(keyword) === -1);
}

function uniquifyFlags(params) {
    params.not = uniq(params.not);
    params.is = uniq(params.is).filter((flag) => params.not.indexOf(flag) === -1);
}

function hasMandatoryParams(params) {
    params = deepCompact(params);

    return mandatoryParams.some((paramName) => params[paramName] != null);
}

// -------------------------------------------------

function parseSearchQuery(query, options) {
    options = Object.assign({ throwOnInvalid: false }, options);

    // Parse query into object
    query = searchQueryParser.parse(query, { keywords: qualifiers });
    query = typeof query === 'string' ? { text: query } : query;
    query = mapKeys(query, (value, key) => camelCase(key)); // CamelCase keys
    delete query.exclude; // We do not use the exclusion feature from search-query-parser
    delete query.offsets; // Remove `offsets` otherwise validation will fail

    // Convert & validate
    const validation = Joi.validate(query, paramsSchema, { abortEarly: options.throwOnInvalid });
    let params = validation.value;

    if (validation.error) {
        // Throw if there's an error and `options.throwOnInvalid` is enabled
        if (options.throwOnInvalid && validation.error) {
            throw validation.error;
        }

        // If `options.throwOnInvalid` is enabled, remove all failed validation props and fill it with defaults
        validation.error.details
        .reverse()
        .forEach((detail) => unset(params, detail.path));

        params = deepCompact(params);
    }

    // Fill in defaults
    params = assignWith(params, paramsDefaults, (tartgetValue, srcValue) => tartgetValue == null ? srcValue : tartgetValue);

    // Post process
    normalizeWeights(params);
    parseKeywords(params);
    uniquifyFlags(params);

    // If no mandatory params were specified (that actually filter results) then we return null
    return hasMandatoryParams(params) ? params : null;
}

function discardQualifiers(query) {
    query = searchQueryParser.parse(query, { keywords: qualifiers });

    return (typeof query === 'string' ? query : query.text || '').trim();
}

module.exports = parseSearchQuery;
module.exports.discardQualifiers = discardQualifiers;
